File: ios/Classes/NimagePlugin.swift
import Flutter
import UIKit

public struct MethodName {
    static let createTexture = "mth_createTexture"
    static let loadImage = "mth_loadImage"
    static let destroyTexture = "mth_destroyTexture"
    static let setVisible = "mth_setVisible"
    static let setInvisible = "mth_setInvisible"
}

public class NimagePlugin: NSObject, FlutterPlugin {
    
    private let textureManager: TextureManager
    
    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "nimage", binaryMessenger: registrar.messenger())
        let instance = NimagePlugin(registrar: registrar)
        registrar.addMethodCallDelegate(instance, channel: channel)
    }
    
    init(registrar: FlutterPluginRegistrar) {
        textureManager = TextureManager(textureRegistry: registrar.textures())
        super.init()
    }
    
    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        let method = call.method
        switch method {
        case MethodName.createTexture:
            createTexture(call, result: result)
        case MethodName.loadImage:
            loadImage(call, result: result)
        case MethodName.destroyTexture:
            destroyTexture(call, result: result)
        case MethodName.setVisible:
            setVisible(call, result: result)
        case MethodName.setInvisible:
            setInvisible(call, result: result)
        default:
            result(FlutterMethodNotImplemented)
        }
    }
    
    private func createTexture(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        let textureId = textureManager.createTexture()
        result(textureId)
    }
    
    private func loadImage(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        if let dict = call.arguments as? Dictionary<String, Any> {
            let textureId = dict["textureId"] as? Int64
            if let imageView = textureManager.getImageTextureView(textureId!) {
                let request = LoadRequest.fromCall(call: call)
                imageView.loadImage(request:request, result: result)
            } else {
                result(FlutterError(code: "NO_TEXTURE", message: "Texture not found", details: nil))
            }
        } else {
            result(FlutterError(code: "INVALID_ARGUMENTS", message: "Invalid arguments", details: nil))
        }
    }
    
    private func destroyTexture(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
          if let textureId = call.arguments as? Int64 {
              textureManager.destroyTexture(textureId)
              result(nil)
          } else {
              result(FlutterError(code: "INVALID_ARGUMENTS", message: "Invalid textureId", details: nil))
          }
      }
      
      private func setVisible(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
          if let textureId = call.arguments as? Int64 {
              if let imageView = textureManager.getImageTextureView(textureId) {
                  imageView.setVisible(true)
                  result(nil)
              } else {
                  result(FlutterError(code: "NO_TEXTURE", message: "Texture not found", details: nil))
              }
          } else {
              result(FlutterError(code: "INVALID_ARGUMENTS", message: "Invalid textureId", details: nil))
          }
      }
      
      private func setInvisible(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
          if let textureId = call.arguments as? Int64 {
              if let imageView = textureManager.getImageTextureView(textureId) {
                  imageView.setVisible(false)
                  result(nil)
              } else {
                  result(FlutterError(code: "NO_TEXTURE", message: "Texture not found", details: nil))
              }
          } else {
              result(FlutterError(code: "INVALID_ARGUMENTS", message: "Invalid textureId", details: nil))
          }
      }
}

File: android/src/main/java/com/flext/nimage/NimagePlugin.java
package com.flext.nimage;

import android.content.Context;

import androidx.annotation.NonNull;
import io.flutter.embedding.engine.plugins.FlutterPlugin;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;

/**
 * NImagePlugin
 */
public class NImagePlugin implements FlutterPlugin, MethodCallHandler {

    private Context context;
    private MethodChannel channel;
    private TextureManager textureManager;
    private static final String MTH_CREATE_TEXTURE = "mth_createTexture";
    private static final String MTH_LOAD_IMAGE = "mth_loadImage";
    private static final String MTH_DESTROY_TEXTURE = "mth_destroyTexture";
    private static final String MTH_SET_VISIBLE = "mth_setVisible";
    private static final String MTH_SET_INVISIBLE = "mth_setInvisible";

    @Override
    public void onAttachedToEngine(@NonNull FlutterPluginBinding flutterPluginBinding) {
        context = flutterPluginBinding.getApplicationContext();
        textureManager = new TextureManager(flutterPluginBinding.getTextureRegistry());
        channel = new MethodChannel(flutterPluginBinding.getBinaryMessenger(), "nimage");
        channel.setMethodCallHandler(this);
    }

    @Override
    public void onMethodCall(@NonNull MethodCall call, @NonNull Result result) {
        if (call.method.equalsIgnoreCase(MTH_CREATE_TEXTURE)) {
            //准备Texture
            long textureId = textureManager.createTexture(context);
            result.success(textureId);
        } else if (call.method.equalsIgnoreCase(MTH_LOAD_IMAGE)) {
            //加载图片
            long textureId = ((Number) call.argument("textureId")).longValue();
            LoadRequest loadRequest = LoadRequest.fromCall(call);
            ImageTextureView imageView = textureManager.getImageTextureView(textureId);
            if (imageView != null) {
                imageView.loadImage(loadRequest, result);
            } else {
                result.error("NO_TEXTURE", "can't find texture with id:" + textureId, null);
            }
        } else if (call.method.equalsIgnoreCase(MTH_SET_VISIBLE)) {
            long textureId = ((Number) call.arguments).longValue();
            ImageTextureView imageView = textureManager.getImageTextureView(textureId);
            if (imageView != null) {
                imageView.setVisible(true);
            }
            result.success(null);
        } else if (call.method.equalsIgnoreCase(MTH_SET_INVISIBLE)) {
            long textureId = ((Number) call.arguments).longValue();
            ImageTextureView imageView = textureManager.getImageTextureView(textureId);
            if (imageView != null) {
                imageView.setVisible(false);
            }
            result.success(null);
        } else if (call.method.equalsIgnoreCase(MTH_DESTROY_TEXTURE)) {
            long textureId = ((Number) call.arguments).longValue();
            textureManager.destroyTexture(textureId);
            result.success(null);
        } else {
            result.notImplemented();
        }
    }

    @Override
    public void onDetachedFromEngine(@NonNull FlutterPluginBinding binding) {
        channel.setMethodCallHandler(null);
    }
}
